Changes for 1674

reshmabidikar opened the pull request:
Includes changes for 1674

Following are the related PRs:

- https://github.com/killbill/killbill-api/pull/114
- https://github.com/killbill/killbill-client-java/pull/117

Note that until the above PRs are merged, tests will not be green (After merging, the correct version needs to be specified in the `killbill/pom.xml`).

pierre commented on catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java:
```diff
@@ -127,30 +119,34 @@ public void uploadCatalog(final String catalogXML, final CallContext callContext
+            return new DefaultCatalogValidation(errors);
         } catch (final JAXBException e) {
-            throw new CatalogApiException(e, ErrorCode.CAT_INVALID_FOR_TENANT, internalTenantContext.getTenantRecordId());
+            errors.add(new ValidationError("Invalid Catalog XML", DefaultVersionedCatalog.class, "")); //TODO_1674 - Error message is hardcoded here, move this elsewhere?
```
See comment below.

pierre commented on catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java:
```diff
@@ -224,5 +220,23 @@ private InternalTenantContext createInternalTenantContext(final TenantContext te
         return internalCallContextFactory.createInternalTenantContextWithoutAccountRecordId(tenantContext);
     }
 
+    private ValidationErrors validateCatalogInternal(final String catalogXML, final InternalTenantContext internalTenantContext) throws ValidationException, JAXBException, CatalogApiException, TransformerException, SAXException, IOException {
```
Instead of throwing here, could you extract error messages from the exception and add them into the errors?

I'm thinking specifically of errors like:

```
				Entity: line 2: parser error : Opening and ending tag mismatch: effectiveDate line 2 and effectiveate
				    <effectiveDate>2013-02-08T00:00:00+00:00</effectiveate>
				                                                           ^
				Entity: line 4: parser error : Opening and ending tag mismatch: recurringBillingMode line 4 and recurringBillingMde
				    <recurringBillingMode>IN_ADVANCE</recurringBillingMde>
				                                                          ^
				Entity: line 10: parser error : Opening and ending tag mismatch: category line 10 and catgory
				            <category>BASE</catgory>
```

or mis-matches in the references (e.g. pricelist with a plan that isn't defined):

```
Caused by: org.xml.sax.SAXParseException; lineNumber: 52; columnNumber: 11; cvc-id.1: There is no ID/IDREF binding for IDREF 'Standard'.
```

reshmabidikar replied:
Fixed as follows:

- added `e.getLinkedException().getMessage()` for `JAXBException` (Since the error message is present in the `linkedException`)
- Added `throw new IllegalStateException(e)` for `TransformerException/IOException/SAXException` to keep this consistent with the earlier `uploadCatalog` code

Let me know how this looks.

sbrossie commented on catalog/src/main/java/org/killbill/billing/catalog/DefaultCatalogValidationError.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public String toString() {
+        final StringBuilder sb = new StringBuilder(errorDescription);
```
Why do we need an extra StringBuilder since we are simply returning the `errorDescription` ?

reshmabidikar replied:
Fixed

sbrossie commented:
> Note that until the above PRs are merged, tests will not be green (After merging, the correct version needs to be specified in the killbill/pom.xml).

@reshmabidikar That should not be the case I think as long as we merge PRs from different repos (`killbill`, `killbill-api`, ...) pertaining to the same underlying issue in a consistent way (i.e at the same time).

So, I would expect: As long as you ensure that your `killbill-api` and `killbill-client-java` and `killbill` repo are all up to date with their respective `work-for-release-0.23.x` branches, then you should be able to use the snapshots generated from such repos and things should work (code should compile as expected).

reshmabidikar commented:
> So, I would expect: As long as you ensure that your killbill-api and killbill-client-java and killbill repo are all up to date with their respective work-for-release-0.23.x branches, then you should be able to use the snapshots generated from such repos and things should work (code should compile as expected).

My `killbill-api` repo is in my [fork](https://github.com/reshmabidikar/killbill-api/tree/work-for-release-0.23.x), so this would not result in a snapshot generation isn't it?

sbrossie commented:
> > So, I would expect: As long as you ensure that your killbill-api and killbill-client-java and killbill repo are all up to date with their respective work-for-release-0.23.x branches, then you should be able to use the snapshots generated from such repos and things should work (code should compile as expected).
> 
> My `killbill-api` repo is in my [fork](https://github.com/reshmabidikar/killbill-api/tree/work-for-release-0.23.x), so this would not result in a snapshot generation isn't it?

I see, you are correct. I did not realize there is no snapshot generation until we merge your `killbill-api` PR. I suppose you can't generate a PR from main `killbill-api` repo, right - you don't have the rights?


If this is the case, what we could do in the future is create such  `killbill-api` PR against another branch `killbill-api:work-for-release-0.23.x-snapshot`, and then merge it early to get the snapshot but it complicates the process.

pierre commented on catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java:
```diff
@@ -127,31 +119,18 @@ public void uploadCatalog(final String catalogXML, final CallContext callContext
+        if (!errors.isEmpty()) {
+            logger.info("Failed to load new catalog version: " + errors.toString());
         }
+        catalogCache.clearCatalog(internalTenantContext);
```
I don't think this is needed?

reshmabidikar replied:
This is required. The [validateCatalogInternal](https://github.com/killbill/killbill/blob/ce2ebe8b05df684f9d6830e81bed8b4791e19f6a/catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java#L206) method [adds](https://github.com/killbill/killbill/blob/ce2ebe8b05df684f9d6830e81bed8b4791e19f6a/catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java#L222) the new catalog to be validated to the existing `VersionedCatalog` in order to perform the validation. This results in the `VersionedCatalog` in the cache being modified during the validation process without an actual catalog upload being performed. So, the cache needs to be cleared after the validation is complete.

pierre replied:
oh, I missed this. This is risky...

Maybe we should clone the catalog first to be safe (using `readExternal`/`writeExternal`). @sbrossie thoughts?

pierre commented on profiles/killbill/src/test/java/org/killbill/billing/jaxrs/TestCatalog.java:
```diff
@@ -292,4 +294,27 @@ public void testGetCatalogJSONWithRecurringBillingMode() throws Exception {
     }
+
+//    @Test(groups = "slow")
+//    public void testValidateCatalog() throws Exception {
```
Cleanup?

reshmabidikar replied:
Had commented out this test since there were some merge issues with the `killbill-client-java` repo which caused compilation errors in this test. The merge issues are now resolved so I've uncommented the test.

pierre commented on catalog/src/main/java/org/killbill/billing/catalog/api/user/DefaultCatalogUserApi.java:
```diff
@@ -224,5 +203,34 @@ private InternalTenantContext createInternalTenantContext(final TenantContext te
+        } catch (final ValidationException e) {
+            errors.addAll(e.getErrors());
+        } catch (final JAXBException e) {
+            errors.add(new ValidationError(e.getLinkedException().getMessage(), DefaultVersionedCatalog.class, ""));
```
Maybe NPE check on `e.getLinkedException()` just in case?

reshmabidikar replied:
Added a null check for `e.getLinkedException()` LMK if this is what you were referring to.

pierre replied:
I don't see it, maybe the PR was merged too early?

reshmabidikar replied:
Yes, the old PR was merged so I've created a new PR: https://github.com/killbill/killbill/pull/1813

pierre reviewed (approved):
A few things otherwise :shipit:

reshmabidikar commented:
@pierre, I've addressed your review comments and created a new PR: https://github.com/killbill/killbill/pull/1813. Please do take another look when you get a chance. Thanks.
